Dump/Load Metabase DB to/from H2 file

ogeagla opened the pull request:
We want to be able to dump an existing Metabase database to a H2 file format, with utilities to be able to encrypt, upload, download, and decrypt the H2 file, so that we can easily chain this to exsiting `load-from-h2` command to have a complete migration.

The way this will work:
 1. Customer purchases hosted instance
 2. Metabase ops target that hosted instance for migration, run initialize methods to generate `s3-upload-url`
 3. Provide `s3-upload-url` to customer so they can run (either through command or API) `dump-to-h2` and `secure-dump-and-upload`
 4. Once customer has successfully run the upload, Metabase ops invokes methods to trigger download and ingest of secure H2 db on the customer's purchased hosted instance.

 - [x] command `dump-to-h2` which takes a source (PG/MySql) db and H2 target filename 
 - [x] command `download-and-unlock` to download and unlock a dump at a public S3 location with provided AES secret
 - [x] command `secure-and-upload` which generates an AES secret and encrypts an H2 dump before it is uploaded to the provided S3 upload url
 - [x] API endpoints for commands: dump-to-h2, secure-and-upload-dump, download-and-unlock-dump, load-from-h2
 - [x] uploads to S3
 - [ ] if Metabase is already running on H2, copy the db over to new filename
 - [ ] override dupe db entries on ingest
 - [ ] compute and save MD5 digest

ogeagla committed:
[cmd] dump to h2 wip

ogeagla committed:
[cmd] dump to h2 wip

ogeagla committed:
[api] wip dump to h2 endpoint

ogeagla committed:
[api] wip dump to h2

camsaul commented:
@ogeagla let me know when this is ready to be reviewed

I think the most important thing for me is that we have some tests that verify everything works as expected end-to-end with a database with some real data in it. The CI steps that test migrating from H2 -> MySQL and H2 -> Postgres are probably good starting points, you can maybe use that same test DB and then run it in reverse and see if we get the same rows in the original H2 DB and the resulting one

ogeagla committed:
[crypto] pub-key for aes secret, aes for dump contents WIP

ogeagla committed:
[crypto] zip/unzip encrypted content

ogeagla committed:
[crypto] wip up/download secure dumps

ogeagla committed:
[s3][test] e2e encryption test, s3 upload/download

ogeagla committed:
[test] crypto resources

ogeagla committed:
[api] secure dump up/down endpoints

ogeagla committed:
[s3] working upload to s3 url

ogeagla committed:
[crypto] decrypt dump from provided secret key from submarine

ogeagla committed:
[secure-dump] wip e2e

ogeagla committed:
[crypto][style] use bytes, break up commands (wip)

ogeagla committed:
[refactor]

ogeagla committed:
[docs]

ogeagla committed:
[docs]

ogeagla committed:
[dump] fix npe when no dump file provided

camsaul reviewed (changes requested):
One thing we'll definitely need to do before merging is make sure we have a good testing story around this

walterl committed:
Ensure that we don't overwrite existing output database

walterl committed:
Cosmetic changes

walterl committed:
Get db type and connection details from specified connection, not global state

walterl committed:
Bail if _output_ db is H2, not input db

walterl committed:
Since the output db value is a filename we can just assume that the db type is H2

ogeagla committed:
[secure-dump] fix async api endpoints

ogeagla committed:
[style]

walterl committed:
Refactor to dump currently configured Metabase db to output H2 dir

ogeagla committed:
[secret] generate aes secret

ogeagla committed:
[secure-dump] rely on user arg for dump path input to secure upload

ogeagla committed:
[style] separate zip ns

ogeagla committed:
[deps] remove amazonica

walterl committed:
Exit with return code, when one was returned

walterl committed:
Upper case column names

H2 (output db) column names are case sensitive and, for some unknown
reason, the H2 db is created with upper cased names, and `cols` are lower-
cased.

walterl committed:
Return positive error code in stead of calling `System/exit` directly

walterl committed:
Output error in red

ogeagla committed:
[api] non-async the dump endpoints
